corehq/motech/forms.py

```diff
@@ -95,8 +94,13 @@ def __init__(self, domain, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
         self.domain = domain
-        self.helper = hqcrispy.HQFormHelper()
-        self.helper.layout = crispy.Layout(
+
+    @property
+    def helper(self):
```

Comment:
Nice. I like this pattern.

Reply:
This is not a caching for performance thing. Every time `self.helper` is accessed it will create and return a new instance, which is quite different from the previous behavior. I can't tell from this code what a `HQFormHelper` does, so can't answer questions like _Is it mutable?_ or _Is it important for things that reference it to always reference the same instance as they did before?_

Reply:
My understanding is that `self.helper` is just part of [Crispy forms](https://django-crispy-forms.readthedocs.io/en/latest/crispy_tag_forms.html#crispy-tag-forms). I got the suggestion from https://stackoverflow.com/questions/13950719/django-crispy-form-init/13954929#13954929. It seems that crispy generally uses it like follows: `{% crispy form form.helper %}`, or, if the helper is stored in `.helper`, it simplifies to `{% crispy form %}`, which is how we're using it: [connection_settings_detail.html](https://github.com/dimagi/commcare-hq/blob/20a84e4e9eb9ab5f308c13ec9fbaba540e0290e6/corehq/motech/templates/motech/connection_settings_detail.html#L14).

According to the crispy docs, creating an 'instance-level' helper is primarily useful over a static, class-based one because it allows us to dynamically modify a helper's attributes. In that case, what I did was functionally different. Changing it as I did would break their example:
```
if redirect_url is not None:
  example_form.helper.form_action = reverse('submit_survey') + '?next=' + redirectUrl
```

That said -- we don't do this. The helper was created once in `__init__`, never modified, and only ever used once in that template. I suppose `cached_property` would guard against errors down the road, if usage changed. On the other hand, I find caching makes code hard to reason about and confusing, and would prefer to avoid it if possible. If you feel strongly about it, I'll add the decorator.

Reply:
The documentation treats it like a mutable object, so having `form.helper` return a new value every time it is accessed is unexpected in this context. I think it would be safer and more intuitive to use `cached_property`.